stl/inc/xutility

```diff
@@ -172,19 +172,28 @@ struct identity {
 template <class _Fx>
 struct _Ref_fn { // pass function object by value as a reference
     template <class... _Args>
-    constexpr decltype(auto) operator()(_Args&&... _Vals) { // forward function call operator
+    constexpr decltype(auto) operator()(_Args&&... _Vals) noexcept(noexcept(_Fn(_STD forward<_Args>(_Vals)...))) {
         return _Fn(_STD forward<_Args>(_Vals)...);
     }
 
     _Fx& _Fn;
 };
 
 template <class _Fn>
-_INLINE_VAR constexpr bool
-    _Pass_functor_by_value_v = sizeof(_Fn) <= sizeof(void*)
-                               && conjunction_v<is_trivially_copy_constructible<_Fn>, is_trivially_destructible<_Fn>>;
+_INLINE_VAR constexpr bool _Pass_functor_by_value_v = conjunction_v<bool_constant<sizeof(_Fn) <= sizeof(void*)>,
+    is_trivially_copy_constructible<_Fn>, is_trivially_destructible<_Fn>>;
 
-template <class _Fn, enable_if_t<_Pass_functor_by_value_v<_Fn>, int> = 0> // TRANSITION, if constexpr
+#if _HAS_IF_CONSTEXPR
+template <class _Fn>
+constexpr auto _Pass_fn(_Fn&& _Val) {
+    if constexpr (_Pass_functor_by_value_v<decay_t<_Fn>>) {
+        return _STD forward<_Fn>(_Val);
+    } else {
+        return _Ref_fn<decay_t<_Fn>>{_Val};

```

Comment:
I'm concerned about an interface change here. Previously, for `!_Pass_functor_by_value_v<_Fn>`, the signature was `_Pass_fn(_Fn&)` taking an lvalue reference, then forming a `_Ref_fn` storing an lvalue reference, which is fine.

Now, `_Pass_fn` accepts rvalues, and when `!_Pass_functor_by_value_v`, is willing to form a `_Ref_fn` ultimately bound to that rvalue. This appears to be tempting doom.

Reply:
I was willing to let this slide given that (1) in permissive mode we allow the lvalue version to accept rvalues anyway thanks to the Evil Extension(TM), and (2) we only use `_Pass_fn` to wrap an expression as an argument to another function, so the lifetime of the return value ends after the same full-expression as would a temporary. (Like `string_view`, there are dangerous uses but this is a safe use.)

Realistically, we only pass lvalue arguments to `_Pass_fn` - which begs the question "Why don't we take the argument by lvalue reference?" I vaguely recall horrible optimizer reactions resulting from binding a reference to a function pointer when @BillyONeal implemented this in the first place. If that's the case, this change needs to be made more carefully and with some codegen investigation to ensure we're not regressing perf.

This is becoming more of a minor work item than a simple drive-by; I'm going to pull the changes to `_Pass_fn` from this PR.